### DISCUSSION TITLE
How to use DevTools when React app is rendered in Shadow DOM?

### DISCUSSION THREAD

**Discussion by DragosMocrii** [TanStack/query#6070] [General]
I have an application that is rendered inside a Shadow DOM. When using React Query Devtools v5, I'm getting an error at line https://github.com/TanStack/query/blob/6e238afae67f7e9a8fee032f5d31598cda1799a3/packages/query-devtools/src/Devtools.tsx#L119 . Ideally, on this line finding the root element should be agnostic to the DOM type used. I don't see any way to define the parent element in the Devtool context either. Any ideas?

**Comment by TkDodo** [comment]
@ardeora FYI

**Comment by ardeora** [comment]
Hello!, Is there a way you could create a small codesandbox/stackblitz reproduction of the issue you're facing with the devtools? I haven't had much experience working with shadow doms before, so a reference point would be immensely helpful!

**Comment by DragosMocrii** [comment]
@ardeora sure, here's an example: https://codesandbox.io/p/sandbox/runtime-fog-32kgtm

**Comment by michelleoj** [comment]
Just bumping this. I'm running into the same issue too and it seems like the issue is that the styles of the devtool (the styles with the id `#_goober`) gets added to the head element of the light dom instead of the shadow dom. Since this library is using goober for its css-in-js functionality, is there any possibility we can update the devtools component to accept a target for [goober to append the style](https://github.com/cristianbote/goober?tab=readme-ov-file#ssr) to an element of our choosing?

**Reply by ardeora** [reply]
This could certainly be exposed as a prop now. We already have a function that creates a custom target to add a nonce to the style element. This function [here](https://github.com/TanStack/query/blob/main/packages/query-devtools/src/utils.tsx#L307-L317) could be expanded to target a specific style tag. Would you be interested in contributing this feature? I would be more than happy to help/assist in anyway. If not, could you please create an issue so I can track this on my backlog? 😄

**Reply by michelleoj** [reply]
Sure, I'd be more than happy to!

**Reply by michelleoj** [reply]
@ardeora Ok opened a PR here https://github.com/TanStack/query/pull/7005
This is my first time contributing to this project so let me know if I missed anything 💃🏾

**Comment by phaethon5882** [comment]
How about using the 👀?
```tsx
export default function Component(...) {
  return (
    <>
       ...
      {createPortal(<ReactQueryDevtools client={queryClient} />, document.body)}
    </>
  );
}
```
